### ISSUE TITLE
Proxy Protocol version 2 Parser has bug

### ISSUE THREAD

**Issue by Charon-Shi** [eclipse/jetty.project#3832]
According to PPv2 protocol:

> PP2_TYPE_NOOP
> The TLV of this type should be ignored when parsed. The value is zero or more
> bytes. Can be used for data padding or alignment. Note that it can be used
> to align only by 3 or more bytes because a TLV can not be smaller than that. 

But In ProxyConnectionFactory.java, next() function in class ProxyProtocolV2Connection, it always read value into bytes array even when length is 1 and value is just one byte, this will break the read workflow.

What's more, the length can be 0 and you need to skip one byte in this situation.

**Comment by gregw** [comment]
@Charon-Shi I don't understand the issue?  The TLVs are read with the following code:
```java 
                        int type = 0xff & _buffer.get();
                        int length = _buffer.getShort();
                        byte[] value = new byte[length];
                        _buffer.get(value);
```
Thus a new array is always created with the correct length of the value.  If the length is 0, the array will be of zero length and no bytes will be consumed.  If the length is 1, the array will be 1 byte long and a single byte will be consumed.

Do you have an example of what you think is badly handled?

**Comment by Charon-Shi** [comment]
If the length is 0, I think you still need 1 byte for value so you can fit 3 bytes as protocol described. Like if the padding bytes are 0x04 0x00 0x00, what Jetty ppv2 will do is to only consume 0x04 0x00 and will choke on next 0x00.

The exception is thrown by _buffer.get(value) and log looks like this:

28 Jun 2019 10:22:35,310 [INFO]  (qtp700366852-25) org.eclipse.jetty.server.ProxyConnectionFactory: Type=4 Length=0 Value= for ProxyV2Connection@5936763c::SocketChannelEndPoint@4b88ad4{/127.0.0.1:65012<->/127.0.0.1:8080,OPEN,fill=-,flush=-,to=5/3600000}{io=0/0,kio=0,kro=1}->ProxyProtocolV2Connection@5936763c
28 Jun 2019 10:22:35,310 [WARN]  (qtp700366852-25) org.eclipse.jetty.server.ProxyConnectionFactory: 
java.lang.ArrayIndexOutOfBoundsException: 0

If the length is 1, exception comes from the same line but looks like this:

....
java.nio.BufferUnderflowException: null
....

**Comment by Charon-Shi** [comment]
Sorry for the ambiguous statement. Let me rephrase:
I think the bug comes from the _buffer.get(value) when the length is 0, which is actually a valid value.

The log looks like this:

```
28 Jun 2019 10:22:35,310 [INFO] (qtp700366852-25) org.eclipse.jetty.server.ProxyConnectionFactory: Type=4 Length=0 Value= for ProxyProtocolV2Connection@5936763c::SocketChannelEndPoint@4b88ad4{/127.0.0.1:65012<->/127.0.0.1:8080,OPEN,fill=-,flush=-,to=5/3600000}{io=0/0,kio=0,kro=1}->ProxyProtocolV2Connection@5936763c
28 Jun 2019 10:22:35,310 [WARN] (qtp700366852-25) org.eclipse.jetty.server.ProxyConnectionFactory: java.lang.ArrayIndexOutOfBoundsException: 0
```

One simple fix is just wrap the _buffer.get() with if statement:
```
if (length > 0) {
    _buffer.get(value);
}
```

I recently try to modify the PPv2 code to let it understand some vendor-specific TLV, I found out the code blob dealing with additional info has a bad code smell, for each fo case, imho, a break statement should be added to avoid cross-case workflow.

**Comment by gregw** [comment]
I've added a test that includes 0 length NOOPs.   I can't see a problem, so closing for now.  Please reopen if you can demonstrate your issue.
